added HttpClient java11 instrumentation for sync requests

kananindzya opened the pull request:
closes #1278 
- [x] sync requests
- [x] async requests
- [x] changelog
- [x] update docs
- [x] test on application and add screenshots(java 11, attach to java7 jvm)

apmmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1307/detail/PR-1307/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1307/detail/PR-1307/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1307/detail/PR-1307/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-java%2Fapm-agent-java-mbp%2FPR-1307/detail/PR-1307/1//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #1307 opened]`

* Reason: `The PR is not allowed to run in the CI yet`

* Start Time: 2020-07-22T17:57:24.691+0000

* Duration: 4 min 52 sec


* Commit: c321d68d0eafb1a4017f9e73eefb5f8de9ebfe86




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: `Error signal`
    * Description: githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permission
    
    * Duration: 0 min 0 sec
    
    * Start Time: 2020-07-22T18:01:13.216+0000
    
    * [log](https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-java/pipelines/apm-agent-java-mbp/pipelines/PR-1307/runs/1/steps/83/log)
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-07-22T18:00:48.827Z] Cleaning workspace
[2020-07-22T18:00:48.846Z] Using shallow fetch with depth 3
[2020-07-22T18:00:48.846Z] Pruning obsolete local branches
[2020-07-22T18:00:48.791Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-java.git # timeout=10
[2020-07-22T18:00:48.797Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2020-07-22T18:00:48.808Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-java.git # timeout=10
[2020-07-22T18:00:48.825Z]  > git rev-parse --verify HEAD # timeout=10
[2020-07-22T18:00:48.831Z] No valid HEAD. Skipping the resetting
[2020-07-22T18:00:48.831Z]  > git clean -fdx # timeout=10
[2020-07-22T18:00:48.846Z] Fetching upstream changes from git@github.com:elastic/apm-agent-java.git
[2020-07-22T18:00:48.847Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-07-22T18:00:48.851Z]  > git fetch --no-tags --progress --prune -- git@github.com:elastic/apm-agent-java.git +refs/pull/1307/head:refs/remotes/origin/PR-1307 +refs/heads/master:refs/remotes/origin/master # timeout=15
[2020-07-22T18:00:49.518Z] Merging remotes/origin/master commit 4de87a80b3daf6c187bfc972ad72dee904ba37fb into PR head commit c321d68d0eafb1a4017f9e73eefb5f8de9ebfe86
[2020-07-22T18:00:50.145Z] Merge succeeded, producing ea357f76e315580c130732d96eaf1b291e37f18f
[2020-07-22T18:00:50.145Z] Checking out Revision ea357f76e315580c130732d96eaf1b291e37f18f (PR-1307)
[2020-07-22T18:00:49.520Z]  > git config core.sparsecheckout # timeout=10
[2020-07-22T18:00:49.528Z]  > git checkout -f c321d68d0eafb1a4017f9e73eefb5f8de9ebfe86 # timeout=15
[2020-07-22T18:00:49.812Z]  > git remote # timeout=10
[2020-07-22T18:00:49.816Z]  > git config --get remote.origin.url # timeout=10
[2020-07-22T18:00:49.824Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-07-22T18:00:49.827Z]  > git merge 4de87a80b3daf6c187bfc972ad72dee904ba37fb # timeout=10
[2020-07-22T18:00:50.124Z]  > git rev-parse HEAD^{commit} # timeout=10
[2020-07-22T18:00:50.143Z]  > git config core.sparsecheckout # timeout=10
[2020-07-22T18:00:50.146Z]  > git checkout -f ea357f76e315580c130732d96eaf1b291e37f18f # timeout=15
[2020-07-22T18:00:53.719Z] Commit message: "Merge commit '4de87a80b3daf6c187bfc972ad72dee904ba37fb' into HEAD"
[2020-07-22T18:00:53.719Z] First time build. Skipping changelog.
[2020-07-22T18:00:53.719Z] Cleaning workspace
[2020-07-22T18:00:53.791Z]  > git --version # timeout=10
[2020-07-22T18:00:53.798Z] fatal: bad object 6c62d3f53b77d98a3e133635021db7fb14d4c2b2
[2020-07-22T18:00:53.716Z]  > git rev-parse --verify HEAD # timeout=10
[2020-07-22T18:00:53.721Z] Resetting working tree
[2020-07-22T18:00:53.721Z]  > git reset --hard # timeout=10
[2020-07-22T18:00:53.741Z]  > git clean -fdx # timeout=10
[2020-07-22T18:00:54.596Z] Masking supported pattern matches of $JOB_GCS_BUCKET or $NOTIFY_TO
[2020-07-22T18:00:54.668Z] Timeout set to expire in 1 hr 0 min
[2020-07-22T18:00:54.696Z] The timestamps step is unnecessary when timestamps are enabled for all Pipeline builds.
[2020-07-22T18:00:54.993Z] [INFO] Number of builds to be searched 10
[2020-07-22T18:00:55.240Z] [INFO] 'shallow' is forced to be disabled when running on PullRequests
[2020-07-22T18:00:55.265Z] Running in /var/lib/jenkins/workspace/_java_apm-agent-java-mbp_PR-1307/src/github.com/elastic/apm-agent-java
[2020-07-22T18:00:55.304Z] [INFO] gitCheckout: Checkout SCM PR-1307 with some customisation.
[2020-07-22T18:00:55.339Z] [INFO] Override default checkout
[2020-07-22T18:00:55.405Z] Sleeping for 10 sec
[2020-07-22T18:01:05.543Z] using credential f6c7695a-671e-4f4f-a331-acdce44ff9ba
[2020-07-22T18:01:05.620Z] Cloning the remote Git repository
[2020-07-22T18:01:05.629Z] Cloning repository git@github.com:elastic/apm-agent-java.git
[2020-07-22T18:01:05.652Z]  > git init /var/lib/jenkins/workspace/_java_apm-agent-java-mbp_PR-1307/src/github.com/elastic/apm-agent-java # timeout=10
[2020-07-22T18:01:05.668Z] Using reference repository: /var/lib/jenkins/.git-references/apm-agent-java.git
[2020-07-22T18:01:05.669Z] Fetching upstream changes from git@github.com:elastic/apm-agent-java.git
[2020-07-22T18:01:05.669Z]  > git --version # timeout=10
[2020-07-22T18:01:05.675Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-07-22T18:01:05.679Z]  > git fetch --tags --progress -- git@github.com:elastic/apm-agent-java.git +refs/heads/*:refs/remotes/origin/* # timeout=10
[2020-07-22T18:01:07.439Z] Checking out Revision c321d68d0eafb1a4017f9e73eefb5f8de9ebfe86 (origin/PR-1307)
[2020-07-22T18:01:06.651Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-java.git # timeout=10
[2020-07-22T18:01:06.655Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2020-07-22T18:01:06.666Z]  > git config remote.origin.url git@github.com:elastic/apm-agent-java.git # timeout=10
[2020-07-22T18:01:06.678Z] Fetching upstream changes from git@github.com:elastic/apm-agent-java.git
[2020-07-22T18:01:06.679Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-07-22T18:01:06.682Z]  > git fetch --tags --progress -- git@github.com:elastic/apm-agent-java.git +refs/pull/1307/head:refs/remotes/origin/PR-1307 +refs/heads/master:refs/remotes/origin/master # timeout=10
[2020-07-22T18:01:07.426Z]  > git rev-parse origin/PR-1307^{commit} # timeout=10
[2020-07-22T18:01:07.436Z]  > git config core.sparsecheckout # timeout=10
[2020-07-22T18:01:07.444Z]  > git checkout -f c321d68d0eafb1a4017f9e73eefb5f8de9ebfe86 # timeout=10
[2020-07-22T18:01:07.621Z] Commit message: "added HttpClient java11 instrumentation for sync requests"
[2020-07-22T18:01:07.621Z] First time build. Skipping changelog.
[2020-07-22T18:01:07.630Z]  > git --version # timeout=10
[2020-07-22T18:01:07.639Z] fatal: bad object 6c62d3f53b77d98a3e133635021db7fb14d4c2b2
[2020-07-22T18:01:08.522Z] Masking supported pattern matches of $GIT_USERNAME or $GIT_PASSWORD
[2020-07-22T18:01:09.113Z] + git fetch https://****:****@github.com/elastic/apm-agent-java.git +refs/pull/*/head:refs/remotes/origin/pr/*
[2020-07-22T18:01:10.105Z] Archiving artifacts
[2020-07-22T18:01:10.799Z] + git rev-parse HEAD
[2020-07-22T18:01:11.163Z] + git rev-parse HEAD
[2020-07-22T18:01:11.474Z] + git rev-parse origin/pr/1307
[2020-07-22T18:01:11.532Z] [INFO] githubEnv: Found Git Build Cause: pr
[2020-07-22T18:01:11.831Z] Masking supported pattern matches of $GITHUB_TOKEN
[2020-07-22T18:01:12.789Z] [WARN] githubApiCall: The REST API call https://api.github.com/repos/elastic/apm-agent-java/pulls/1307/reviews return 0 elements
[2020-07-22T18:01:12.864Z] [INFO] githubPrCheckApproved: Title: wip: added HttpClient java11 instrumentation for sync requests - User: kananindzya - Author Association: CONTRIBUTOR
[2020-07-22T18:01:13.234Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet
[2020-07-22T18:01:13.234Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permissions can do so.)
[2020-07-22T18:01:13.321Z] [INFO] Let's stop build #1. The PR is not allowed to run in the CI yet
[2020-07-22T18:01:13.346Z] Sleeping for 5 sec
[2020-07-22T18:01:14.335Z] Stage "Build" skipped due to earlier failure(s)
[2020-07-22T18:01:14.520Z] Stage "Tests" skipped due to earlier failure(s)
[2020-07-22T18:01:14.705Z] Stage "Unit Tests" skipped due to earlier failure(s)
[2020-07-22T18:01:14.717Z] Stage "Smoke Tests 01" skipped due to earlier failure(s)
[2020-07-22T18:01:14.728Z] Stage "Smoke Tests 02" skipped due to earlier failure(s)
[2020-07-22T18:01:14.738Z] Stage "Benchmarks" skipped due to earlier failure(s)
[2020-07-22T18:01:14.750Z] Stage "Javadoc" skipped due to earlier failure(s)
[2020-07-22T18:01:14.906Z] Failed in branch Unit Tests
[2020-07-22T18:01:14.915Z] Failed in branch Smoke Tests 01
[2020-07-22T18:01:14.923Z] Failed in branch Smoke Tests 02
[2020-07-22T18:01:14.931Z] Failed in branch Benchmarks
[2020-07-22T18:01:14.941Z] Failed in branch Javadoc
[2020-07-22T18:01:15.182Z] Stage "Integration Tests" skipped due to earlier failure(s)
[2020-07-22T18:01:15.273Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2020-07-22T18:01:15.313Z] Stage "AfterRelease" skipped due to earlier failure(s)
[2020-07-22T18:01:15.613Z] Running on Jenkins in /var/lib/jenkins/workspace/_java_apm-agent-java-mbp_PR-1307
[2020-07-22T18:01:15.764Z] [INFO] getVaultSecret: Getting secrets
[2020-07-22T18:01:15.862Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-07-22T18:01:16.731Z] + chmod 755 generate-build-data.sh
[2020-07-22T18:01:16.731Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1307/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1307/runs/1 ABORTED 231771
[2020-07-22T18:01:16.982Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-java/apm-agent-java-mbp/PR-1307/runs/1/steps/?limit=10000 -o steps-info.json

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

kananindzya commented:
Hi @felixbarny ,
I need advice to instrument asyncMethod
```
HttpClient.sendAsync(request, HttpResponse.BodyHandlers.ofString()).get();    
```
returns CompletableFuture. I am confuse what to do with it )

felixbarny commented:
You can register a callback by calling `future.whenComplete`. In that callback, you can end the span. Note that with indy plugins, you don't have to duplicate code in the advices anymore. You can just share common code between `HttpClientAsyncInstrumentation` and `HttpClientInstrumentation` in a helper class. You don't need a helper class manager for that and you can use framework-specific types everywhere in an indy plugin, not just the advice methods.
Also, you can return `HttpRequest` instead of `Object` and get rid of the `typing = Assigner.Typing.DYNAMIC`.

kananindzya commented:
> You can register a callback by calling `future.whenComplete`. In that callback, you can end the span. Note that with indy plugins, you don't have to duplicate code in the advices anymore. You can just share common code between `HttpClientAsyncInstrumentation` and `HttpClientInstrumentation` in a helper class. You don't need a helper class manager for that and you can use framework-specific types everywhere in an indy plugin, not just the advice methods.
> Also, you can return `HttpRequest` instead of `Object` and get rid of the `typing = Assigner.Typing.DYNAMIC`.

ok, thanks. I will try to implement this today

felixbarny commented:
Rather than cloning the whole `HttpRequest` to add the trace headers, you could just instrument `HttpRequest#headers()` and return a copy of the `HttpHeaders` that include the trace headers.
Another simplification you may want to implement is to remove `GlobalThreadLocal<Span>` and call `tracer.getActiveExitSpan()` on the exit advice instead.

kananindzya commented:
> Rather than cloning the whole `HttpRequest` to add the trace headers, you could just instrument `HttpRequest#headers()` and return a copy of the `HttpHeaders` that include the trace headers.
> Another simplification you may want to implement is to remove `GlobalThreadLocal<Span>` and call `tracer.getActiveExitSpan()` on the exit advice instead.

added HttpRequest#headers instrumentation, removed GlobalThreadLocal<Span>

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/pom.xml:
```diff
@@ -0,0 +1,46 @@
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>11</source>
+                    <target>11</target>
```
Trying to load the instrumentations of this module would fail of the agent is attached to a Java 7 JVM, for example. But maybe that can easily be fixed by just catching the `java.lang.UnsupportedClassVersionError` in `co.elastic.apm.agent.util.DependencyInjectingServiceLoader#instantiate(java.lang.ClassLoader, java.lang.String)`. That would be quite nice.

kananindzya replied:
I added exception handling to `DependencyInjectingServiceLoader#instantiate`

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientAsyncInstrumentation.java:
```diff
@@ -0,0 +1,155 @@
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return named("jdk.internal.net.http.HttpClientImpl");
```
don't depend on the exact internal name. instead, use `hasSuperType(named("java.net.http.HttpClient"))` and add a pre matcher with `nameContains("HttpClient")`

kananindzya replied:
fixed

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientAsyncInstrumentation.java:
```diff
@@ -0,0 +1,155 @@
+            if (span != null) {
+                spanTls.set(span);
+                span.activate();
+                HttpRequest.Builder builder = HttpRequest.newBuilder(httpRequest.uri())
```
rely on the `HttpRequest#header` instrumentation here as well instead of cloning the request

kananindzya replied:
sorry, I should have deleted this instrumentation. Now I am working on it

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientAsyncInstrumentation.java:
```diff
@@ -0,0 +1,155 @@
+    public static class HttpClient11Advice {
+
+        @VisibleForAdvice
+        public final static GlobalThreadLocal<Span> spanTls = GlobalThreadLocal.get(HttpClient11Advice.class, "spanTls");
```
replace with `tracer.getActiveExitSpan()`

kananindzya replied:
replaced

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientInstrumentation.java:
```diff
@@ -0,0 +1,73 @@
+            URI uri = httpRequest.uri();
+            co.elastic.apm.agent.http.client.HttpClientHelper.startHttpClientSpan(parent, httpRequest.method(), uri.toString(), uri.getScheme(),
+                HttpClientHelper.computeHostName(uri.getHost()), uri.getPort())
+                .activate();
```
check for null

kananindzya replied:
added check

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientHelper.java:
```diff
@@ -0,0 +1,22 @@
+import javax.annotation.Nullable;
+
+@VisibleForAdvice
+public class HttpClientHelper {
```
Rename so the name doesn't clash with `co.elastic.apm.agent.http.client.HttpClientHelper`

kananindzya replied:
renamed

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientHelper.java:
```diff
@@ -0,0 +1,22 @@
+
+import javax.annotation.Nullable;
+
+@VisibleForAdvice
```
When using indy plugins, this annotation is not needed. You can even make the class package-private if you want to.

kananindzya replied:
deleted

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpRequestHeadersInstrumentation.java:
```diff
@@ -0,0 +1,64 @@
+            return HttpHeaders.of(headersMap, (x, y) -> true);
+        }
+    }
+
```
add pre-matcher to speed up matching `nameContains("HttpClient")`

kananindzya replied:
added pre matcher filter

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/resources/META-INF/services/co.elastic.apm.agent.sdk.ElasticApmInstrumentation:
```diff
@@ -0,0 +1,2 @@
+co.elastic.apm.agent.httpclient.HttpClientInstrumentation
+co.elastic.apm.agent.httpclient.HttpRequestHeadersInstrumentation
```
```suggestion
co.elastic.apm.agent.httpclient.HttpClientAsyncInstrumentation
co.elastic.apm.agent.httpclient.HttpRequestHeadersInstrumentation
```

kananindzya replied:
added HttpClientAsyncInstrumentation

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientAsyncInstrumentation.java:
```diff
@@ -0,0 +1,155 @@
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class, inline = false)
+        public static void onAfterExecute(@Advice.Return @Nullable HttpResponse response,
```
the return value is a `CompletableFuture`, not a `HttpResponse`.

kananindzya replied:
fixed

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientAsyncInstrumentation.java:
```diff
@@ -0,0 +1,155 @@
+        return true;
+    }
+
+    public abstract static class AbstractCompletableFutureInstrumentation extends HttpClientAsyncInstrumentation {
```
You don't have to instrument the CompletableFuture. Just call future.whenComplete

kananindzya replied:
added calling future.whenComplete

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientAsyncInstrumentation.java:
```diff
@@ -0,0 +1,80 @@
+                return;
+            }
+            activeSpan.deactivate();
+            AsyncCallbackCreator callbackCreator = new AsyncCallbackCreatorImpl();
```
These creator interfaces are not needed anymore with indy plugins :)
Just directly implement the BiConsumer as a lambda.

kananindzya replied:
fixed

felixbarny commented on apm-agent-core/src/main/java/co/elastic/apm/agent/util/DependencyInjectingServiceLoader.java:
```diff
@@ -129,6 +129,9 @@ private T instantiate(ClassLoader classLoader, String implementation) {
             throw new ServiceConfigurationError(msg, e);
+        } catch(UnsupportedClassVersionError e) {
+            String msg = String.format("unable to instantiate '%s', unsupported class version error", implementation);
+            throw new ServiceConfigurationError(msg, e);
```
Doesn’t that propagate the error and crash the agent? Or is it caught somewhere else?

kananindzya replied:
No, it's not caught anywhere else. Here should I just catch and ignore it?

felixbarny replied:
Maybe just log your `msg` on debug. But I'm not sure if that actually works out. Please also try the agent with an older Java version to verify that it doesn't crash.

kananindzya replied:
I run on on Java 1.8.0_242 via javaagent and got exception
```
2020-08-14 03:02:47,266 [main] DEBUG co.elastic.apm.agent.configuration.StartupInfo - service_name: 'spring-hello-service' (source: elasticapm.properties)
2020-08-14 03:02:47,266 [main] DEBUG co.elastic.apm.agent.configuration.StartupInfo - disable_instrumentations: '' (source: elasticapm.properties)
2020-08-14 03:02:47,266 [main] DEBUG co.elastic.apm.agent.configuration.StartupInfo - application_packages: 'com.kananinja.apm.spring' (source: elasticapm.properties)
2020-08-14 03:02:47,267 [main] DEBUG co.elastic.apm.agent.configuration.StartupInfo - log_level: 'DEBUG' (source: elasticapm.properties)
2020-08-14 03:02:47,267 [main] DEBUG co.elastic.apm.agent.configuration.StartupInfo - enable_log_correlation: 'true' (source: elasticapm.properties)
2020-08-14 03:02:47,268 [main] DEBUG co.elastic.apm.agent.bci.ElasticApmAgent - No plugins dir
Failed to start agent
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at co.elastic.apm.agent.bci.AgentMain.init(AgentMain.java:103)
	at co.elastic.apm.agent.bci.AgentMain.premain(AgentMain.java:57)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386)
	at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401)
Caused by: java.util.ServiceConfigurationError: unable to instantiate 'co.elastic.apm.agent.httpclient.HttpClientInstrumentation', unsupported class version error
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.instantiate(DependencyInjectingServiceLoader.java:134)
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.instantiate(DependencyInjectingServiceLoader.java:112)
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.<init>(DependencyInjectingServiceLoader.java:66)
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.load(DependencyInjectingServiceLoader.java:86)
	at co.elastic.apm.agent.bci.ElasticApmAgent.loadInstrumentations(ElasticApmAgent.java:158)
	at co.elastic.apm.agent.bci.ElasticApmAgent.initInstrumentation(ElasticApmAgent.java:150)
	at co.elastic.apm.agent.bci.ElasticApmAgent.initialize(ElasticApmAgent.java:136)
	... 12 more
Caused by: java.lang.UnsupportedClassVersionError: co/elastic/apm/agent/httpclient/HttpClientInstrumentation has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.instantiate(DependencyInjectingServiceLoader.java:118)
	... 18 more
```

kananindzya replied:
log msg with error level
```
2020-08-14 03:11:12,918 [main] DEBUG co.elastic.apm.agent.configuration.StartupInfo - enable_log_correlation: 'true' (source: elasticapm.properties)
2020-08-14 03:11:12,920 [main] DEBUG co.elastic.apm.agent.bci.ElasticApmAgent - No plugins dir
2020-08-14 03:11:12,942 [main] ERROR co.elastic.apm.agent.util.DependencyInjectingServiceLoader - unable to instantiate 'co.elastic.apm.agent.httpclient.HttpClientInstrumentation', unsupported class version error
java.lang.UnsupportedClassVersionError: co/elastic/apm/agent/httpclient/HttpClientInstrumentation has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
	at java.lang.Class.forName0(Native Method) ~[?:1.8.0_242]
	at java.lang.Class.forName(Class.java:348) ~[?:1.8.0_242]
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.instantiate(DependencyInjectingServiceLoader.java:126) [?:?]
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.instantiate(DependencyInjectingServiceLoader.java:116) [?:?]
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.<init>(DependencyInjectingServiceLoader.java:70) [?:?]
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.load(DependencyInjectingServiceLoader.java:90) [?:?]
	at co.elastic.apm.agent.bci.ElasticApmAgent.loadInstrumentations(ElasticApmAgent.java:158) [?:?]
	at co.elastic.apm.agent.bci.ElasticApmAgent.initInstrumentation(ElasticApmAgent.java:150) [?:?]
	at co.elastic.apm.agent.bci.ElasticApmAgent.initialize(ElasticApmAgent.java:136) [?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_242]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_242]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_242]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_242]
	at co.elastic.apm.agent.bci.AgentMain.init(AgentMain.java:103) [elastic-apm-agent-1.18.1.RC1-SNAPSHOT.jar:?]
	at co.elastic.apm.agent.bci.AgentMain.premain(AgentMain.java:57) [elastic-apm-agent-1.18.1.RC1-SNAPSHOT.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_242]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_242]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_242]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_242]
	at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386) [?:1.8.0_242]
	at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401) [?:1.8.0_242]
2020-08-14 03:11:12,949 [main] ERROR co.elastic.apm.agent.util.DependencyInjectingServiceLoader - unable to instantiate 'co.elastic.apm.agent.httpclient.HttpClientAsyncInstrumentation', unsupported class version error
java.lang.UnsupportedClassVersionError: co/elastic/apm/agent/httpclient/HttpClientAsyncInstrumentation has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
	at java.lang.Class.forName0(Native Method) ~[?:1.8.0_242]
	at java.lang.Class.forName(Class.java:348) ~[?:1.8.0_242]
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.instantiate(DependencyInjectingServiceLoader.java:126) [?:?]
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.instantiate(DependencyInjectingServiceLoader.java:116) [?:?]
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.<init>(DependencyInjectingServiceLoader.java:70) [?:?]
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.load(DependencyInjectingServiceLoader.java:90) [?:?]
	at co.elastic.apm.agent.bci.ElasticApmAgent.loadInstrumentations(ElasticApmAgent.java:158) [?:?]
	at co.elastic.apm.agent.bci.ElasticApmAgent.initInstrumentation(ElasticApmAgent.java:150) [?:?]
	at co.elastic.apm.agent.bci.ElasticApmAgent.initialize(ElasticApmAgent.java:136) [?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_242]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_242]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_242]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_242]
	at co.elastic.apm.agent.bci.AgentMain.init(AgentMain.java:103) [elastic-apm-agent-1.18.1.RC1-SNAPSHOT.jar:?]
	at co.elastic.apm.agent.bci.AgentMain.premain(AgentMain.java:57) [elastic-apm-agent-1.18.1.RC1-SNAPSHOT.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_242]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_242]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_242]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_242]
	at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386) [?:1.8.0_242]
	at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401) [?:1.8.0_242]
2020-08-14 03:11:12,954 [main] ERROR co.elastic.apm.agent.util.DependencyInjectingServiceLoader - unable to instantiate 'co.elastic.apm.agent.httpclient.HttpRequestHeadersInstrumentation', unsupported class version error
java.lang.UnsupportedClassVersionError: co/elastic/apm/agent/httpclient/HttpRequestHeadersInstrumentation has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
	at java.lang.Class.forName0(Native Method) ~[?:1.8.0_242]
	at java.lang.Class.forName(Class.java:348) ~[?:1.8.0_242]
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.instantiate(DependencyInjectingServiceLoader.java:126) [?:?]
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.instantiate(DependencyInjectingServiceLoader.java:116) [?:?]
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.<init>(DependencyInjectingServiceLoader.java:70) [?:?]
	at co.elastic.apm.agent.util.DependencyInjectingServiceLoader.load(DependencyInjectingServiceLoader.java:90) [?:?]
	at co.elastic.apm.agent.bci.ElasticApmAgent.loadInstrumentations(ElasticApmAgent.java:158) [?:?]
	at co.elastic.apm.agent.bci.ElasticApmAgent.initInstrumentation(ElasticApmAgent.java:150) [?:?]
	at co.elastic.apm.agent.bci.ElasticApmAgent.initialize(ElasticApmAgent.java:136) [?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_242]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_242]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_242]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_242]
	at co.elastic.apm.agent.bci.AgentMain.init(AgentMain.java:103) [elastic-apm-agent-1.18.1.RC1-SNAPSHOT.jar:?]
	at co.elastic.apm.agent.bci.AgentMain.premain(AgentMain.java:57) [elastic-apm-agent-1.18.1.RC1-SNAPSHOT.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_242]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_242]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_242]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_242]
	at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386) [?:1.8.0_242]
	at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401) [?:1.8.0_242]
2020-08-14 03:11:13,014 [main] DEBUG co.elastic.apm.agent.bci.ElasticApmAgent - Applying instrumentation co.elastic.apm.agent.httpclient.ApacheHttpClientInstrumentation
2020-08-14 03:11:13,022 [main] DEBUG co.elastic.apm.agent.bci.ElasticApmAgent - Applying instrumentation co.elastic.apm.agent.httpclient.ApacheHttpAsyncClientInstrumentation
2020-08-14 03:11:13,023 [main] DEBUG co.elastic.apm.agent.bci.ElasticApmAgent - Applying instrumentation co.elastic.apm.ag
```

felixbarny replied:
So it did work out to catch the error?

I'd just log the message `String.format("unable to instantiate '%s', unsupported class version error: %s", implementation, e.getMessage())` on `DEBUG` log, not print the whole stack trace on `ERROR`. That's because this exception is expected when running the agent on an older JVM.

kananindzya replied:
Yes, error catched and ignored:
```
2020-08-16 02:12:33,350 [main] DEBUG co.elastic.apm.agent.configuration.StartupInfo - log_level: 'DEBUG' (source: Java System Properties)
2020-08-16 02:12:33,351 [main] DEBUG co.elastic.apm.agent.bci.ElasticApmAgent - No plugins dir
2020-08-16 02:12:33,376 [main] ERROR co.elastic.apm.agent.util.DependencyInjectingServiceLoader - unable to instantiate 'co.elastic.apm.agent.httpclient.HttpClientInstrumentation', unsupported class version error: co/elastic/apm/agent/httpclient/HttpClientInstrumentation has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
2020-08-16 02:12:33,376 [main] ERROR co.elastic.apm.agent.util.DependencyInjectingServiceLoader - unable to instantiate 'co.elastic.apm.agent.httpclient.HttpClientAsyncInstrumentation', unsupported class version error: co/elastic/apm/agent/httpclient/HttpClientAsyncInstrumentation has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
2020-08-16 02:12:33,377 [main] ERROR co.elastic.apm.agent.util.DependencyInjectingServiceLoader - unable to instantiate 'co.elastic.apm.agent.httpclient.HttpRequestHeadersInstrumentation', unsupported class version error: co/elastic/apm/agent/httpclient/HttpRequestHeadersInstrumentation has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0
2020-08-16 02:12:33,431 [main] DEBUG co.elastic.apm.agent.bci.ElasticApmAgent - Applying instrumentation co.elastic.apm.agent.httpclient.ApacheHttpClientInstrumentation
2020-08-16 02:12:33,439 [main] DEBUG co.elastic.apm.agent.bci.ElasticApmAgent - Applying instrumentation co.elastic.apm.agent.httpclient.ApacheHttpAsyncClientInstrumentation
2020-08-16 02:12:33,440 [main] DEBUG co.elastic.apm.agent.bci.ElasticApmAgent - Applying instrumentation co.elastic.apm.agent.httpclient.ApacheHttpAsyncClientRedirectInstrumentati
```

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/test/java/co/elastic/apm/agent/httpclient/HttpClientAsyncInstrumentationTest.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Override
+    public void assertCircularRedirect() {
+        assertThat(reporter.getSpans().get(0).getContext().getHttp().getStatusCode()).isEqualTo(303);
```
Why is this necessary? Doesn't this fail on circular redirects? If so, override the test method to just skip it.

kananindzya replied:
Yes, this fail on circluar redirects.
Ok, I overrided this test method.

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpRequestHeadersInstrumentation.java:
```diff
@@ -0,0 +1,95 @@
+    public static class HttpRequestHeadersAdvice {
+
+        @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+        public static void onBeforeExecute() {
```
Can this be removed?

kananindzya replied:
Removed

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientInstrumentation.java:
```diff
@@ -0,0 +1,107 @@
+        public static void onAfterExecute(@Advice.Return @Nullable HttpResponse response,
+                                          @Advice.Thrown @Nullable Throwable t) {
+            final Span span = tracer.getActiveExitSpan();
+            if (span != null) {
```
nit: better return here to reduce the level of nesting

kananindzya replied:
fixed

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientAsyncInstrumentation.java:
```diff
@@ -0,0 +1,113 @@
+                return;
+            }
+            activeSpan.deactivate();
+            BiConsumer<HttpResponse, Throwable> callback = (response, throwable) -> {
+                try {
+                    if (response != null) {
+                        int statusCode = response.statusCode();
+                        activeSpan.getContext().getHttp().withStatusCode(statusCode);
+                    }
+                    activeSpan.captureException(throwable);
+                } finally {
+                    activeSpan.end();
+                }
+            };
+            if (completableFuture != null) {
+                completableFuture.whenComplete(callback);
```
```suggestion
            if (completableFuture != null) {
                completableFuture.whenComplete((response, throwable) -> {
                    try {
                        if (response != null) {
                            int statusCode = response.statusCode();
                            activeSpan.getContext().getHttp().withStatusCode(statusCode);
                        }
                        activeSpan.captureException(throwable);
                    } finally {
                        activeSpan.end();
                    }
                });
```

kananindzya commented:
screenshots
![image](https://user-images.githubusercontent.com/48118512/90320929-04994300-df67-11ea-954a-efca3bb8ce9e.png)
![image](https://user-images.githubusercontent.com/48118512/90320939-1f6bb780-df67-11ea-846e-c8aff4f9f783.png)

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClient11Helper.java:
```diff
@@ -0,0 +1,45 @@
+        if (originalHostName != null && originalHostName.contains(":") && !originalHostName.startsWith("[")) {
+            StringBuilder sb = new StringBuilder();
+            sb.setLength(0);
+            sb.append("[").append(originalHostName).append("]");
```
why is this necessary?

kananindzya replied:
deleted

felixbarny commented on apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientAsyncInstrumentation.java:
```diff
@@ -0,0 +1,112 @@
+            }
+            final AbstractSpan<?> parent = tracer.getActive();
+            URI uri = httpRequest.uri();
+            Span span = HttpClientHelper.startHttpClientSpan(parent, httpRequest.method(), uri.toString(), uri.getScheme(),
```
Would it be possible to use `co.elastic.apm.agent.http.client.HttpClientHelper#startHttpClientSpan(co.elastic.apm.agent.impl.transaction.AbstractSpan<?>, java.lang.String, java.net.URI, java.lang.CharSequence)` i.e. `HttpClientHelper.startHttpClientSpan(parent, httpRequest.method(), uri, null)`?

Or are you using the other variant for a particular reason?

kananindzya replied:
replaced

SylvainJuge commented:
Hi @kananindzya ,

I've pushed some changes in a PR on your branch (https://github.com/kananindzya/apm-agent-java/pull/7), as it was easier than explaining them.

That's mostly cosmetic changes, this new plugin is really great :rocket: .

kananindzya commented:
> Hi @kananindzya ,
> 
> I've pushed some changes in a PR on your branch ([kananindzya#7](https://github.com/kananindzya/apm-agent-java/pull/7)), as it was easier than explaining them.
> 
> That's mostly cosmetic changes, this new plugin is really great rocket .

Merged. Thanks you :)

felixbarny commented on apm-agent-plugins/apm-jdk-httpclient-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientAsyncInstrumentation.java:
```diff
@@ -0,0 +1,96 @@
+
+    @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+    public static void onBeforeExecute(@Advice.Argument(value = 0) HttpRequest httpRequest) {
+        startSpan(httpRequest);
```
[very minor] potentially return the span here (as an Object) and add a parameter `@Nullable @Advice.Enter Object spanObj` on the exit advice. This can replace the `final Span span = tracer.getActiveExitSpan();` statement in the exit advice.

The same can be done in `HttpClientInstrumentation`

kananindzya replied:
done

felixbarny commented:
Javadoc fails:
```

[2020-08-26T08:12:25.703Z] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.1.1:javadoc (default-cli) on project apm-agent-plugin-sdk: An error has occurred in Javadoc report generation: 
[2020-08-26T08:12:25.703Z] [ERROR] Exit code: 1 - javadoc: error - The code being documented uses packages in the unnamed module, but the packages defined in https://docs.oracle.com/en/java/javase/11/docs/api/ are in named modules.
```

Maybe related to https://github.com/elastic/apm-agent-java/pull/1237?

SylvainJuge commented:
That might be a side effect of compiling test code to Java 11 (from 10 initially), I'll investigate that.

subeenn commented:
@SylvainJuge @kananindzya,  in which agent version this will be available ?

SylvainJuge commented:
It has been merged 3 weeks ago and is included in the latest available agent `1.18.0`
